### PR TITLE
#2219 Complete setting definition of docs

### PR DESCRIPTION
**Pull request by saivirtue**
### Checklist
- [x] [Review Contribution Guidelines](https://github.com/oharastream/ohara/blob/master/CONTRIBUTING.md)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#authors))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #2219)
- [ ] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: docs/src/custom_streamapp.rst [hunk 1]
```diff
@@ -87,28 +87,25 @@ A base implementation for a custom streamApp only need to include
...
+     public void start(OStream<Row> ostream, StreamDefinitions streamDefinitions) {
        ostream.start();
      }
    }
 
-.. _streamapp-init-method:
-
 .. note::
-   The methods we provide here belong to Ohara StreamApp, which have
+   The following methods we provided belong to Ohara StreamApp, which have
```
Comment (konekoya) [inline docs/src/custom_streamapp.rst]:
```suggestion
   The following methods we provided belongs to Ohara StreamApp, which has
```

File: docs/src/custom_streamapp.rst [hunk 2]
```diff
@@ -118,36 +115,85 @@ StreamApp is **init()**. This is an optional method that can be used for
 user to initialize some external data source connections or input
 parameters.
 
+.. _streamapp-config-method:
+
+config() method
+~~~~~~~~~~~~~~~
+
+In a streamApp application, you may want to config your own parameters. We support a method here to help you define
+a custom streamDefinitions list in streamApp. The details of streamDefinitions is list :ref:`here <streamapp-setting-definitions>`.
```
Comment (konekoya) [inline docs/src/custom_streamapp.rst]:
```suggestion
a custom streamDefinitions list in streamApp. The details of streamDefinitions are list :ref:`here <streamapp-setting-definitions>`.
```

File: docs/src/custom_streamapp.rst [hunk 3]
```diff
@@ -253,11 +299,46 @@ in javadoc.
...
 .. _streamapp-setting-definitions:
 
-Setting Definitions
--------------------
+StreamApp Definitions
+---------------------
 
-Will be implemented in the near future. Also see issue: :ohara-issue:`962`
+StreamApp stores a list of :ref:`SettingDef <setting-definition>`, which is StreamDefinitions, in the data store.
+By default, we will keep the following definitions in "core" group and generate the definition in stream api :
```
Comment (konekoya) [inline docs/src/custom_streamapp.rst]:
```suggestion
By default, we will keep the following definitions in the "core" group and generate the definition in stream API :
```

File: docs/src/custom_streamapp.rst [hunk 4]
```diff
@@ -253,11 +299,46 @@ in javadoc.
...
+
+#. DefaultConfigs.BROKER_DEFINITION : The broker list
+#. DefaultConfigs.IMAGE_NAME_DEFINITION : The image name
+#. DefaultConfigs.NAME_DEFINITION : The streamApp application name
+#. DefaultConfigs.GROUP_DEFINITION : The streamApp group name
+#. DefaultConfigs.FROM_TOPICS_DEFINITION : The from topic
+#. DefaultConfigs.TO_TOPICS_DEFINITION : The to topic
+#. DefaultConfigs.JMX_PORT_DEFINITION : The expose jmx port
```
Comment (konekoya) [inline docs/src/custom_streamapp.rst]:
```suggestion
#. DefaultConfigs.JMX_PORT_DEFINITION : The exposed jmx port
```

File: docs/src/custom_streamapp.rst [hunk 5]
```diff
@@ -253,11 +299,46 @@ in javadoc.
...
+#. DefaultConfigs.INSTANCES_DEFINITION : The running instances
+#. DefaultConfigs.NODE_NAMES_DEFINITION : The node name list
+#. DefaultConfigs.VERSION_DEFINITION : The version of streamApp
+#. DefaultConfigs.REVISION_DEFINITION : The revision of streamApp
+#. DefaultConfigs.AUTHOR_DEFINITION : The author of streamApp
+#. DefaultConfigs.TAGS_DEFINITION : The tags of streamApp
+
+Any other definition except above list will be trick as custom definition. You can define
```
Comment (konekoya) [inline docs/src/custom_streamapp.rst]:
```suggestion
Any other definition except above list will be treated as a custom definition. You can define
```

File: docs/src/custom_streamapp.rst [hunk 6]
```diff
@@ -253,11 +299,46 @@ in javadoc.
...
+
+.. code-block:: java
+
+   SettingDef.builder().key(joinTopic).group("default").build()
+
+as a definition that is list in "default" group, or
```
Comment (konekoya) [inline docs/src/custom_streamapp.rst]:
```suggestion
as a definition that is listed in "default" group, or
```

File: docs/src/custom_streamapp.rst [hunk 7]
```diff
@@ -253,11 +299,46 @@ in javadoc.
...
+
+.. code-block:: java
+
+   SettingDef.builder().key(otherKey).group("common").build()
+
+as a definition that is list in "common" group.
```
Comment (konekoya) [inline docs/src/custom_streamapp.rst]:
```suggestion
as a definition that is listed in the "common" group.
```

File: docs/src/custom_streamapp.rst [hunk 8]
```diff
@@ -118,36 +115,85 @@ StreamApp is **init()**. This is an optional method that can be used for
 user to initialize some external data source connections or input
 parameters.
 
+.. _streamapp-config-method:
+
+config() method
+~~~~~~~~~~~~~~~
+
+In a streamApp application, you may want to config your own parameters. We support a method here to help you define
```
Comment (konekoya) [inline docs/src/custom_streamapp.rst]:
```suggestion
In a streamApp application, you may want to configure your own parameters. We support a method here to help you define
```

File: docs/src/setting_definition.rst [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+.. limitations under the License.
+..
+
+.. _setting-definition:
+
+Setting Definition Guide
+========================
+
+A powerful application always has a complicated configuration. In order to
+be a best friend of ohara users, ohara provides a method which
```
Comment (vitojeng) [inline docs/src/setting_definition.rst]: Let us use `Ohara` instead `ohara`. 😄 
```suggestion
be a best friend of Ohara users, Ohara provides a method which
```

File: docs/src/setting_definition.rst [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+----------
+
+In a custom application, the settings could have various data type. In order to display correct data type in
+ohara manager and leverage the benefit of :ref:`type checker<checker>`, we strongly suggest you to define the
+correct data type for each setting.
+
+The following data types are supported currently.
+
+Type.BOOLEAN
+''''''''''''
```
Comment (vitojeng) [inline docs/src/setting_definition.rst]: Let us use the following character for the header:
```
 Level 1
 ===========

 Level 2
 -----------

 Level 3
 ^^^^^^^^^^^

 Level 4
 """""""""""
```

File: docs/src/setting_definition.rst (hunk 9 above)
Comment (chia7712) [reply docs/src/setting_definition.rst]: Nice review. We should correct this important keyword - Ohara - for whole project :)

Comment (vitojeng) [reply docs/src/setting_definition.rst]: Please also check the whole file. ^^

File: docs/src/setting_definition.rst (hunk 10 above)
Comment (saivirtue) [reply docs/src/setting_definition.rst]: 👍

**Comment by chia7712** [comment]
retry

**Comment by saivirtue** [comment]
retry

**Pull request merged**